### PR TITLE
Add high-level `select` and `selectStatus` functions

### PR DESCRIPTION
**Pull request by hallettj**
I love this library! After applying it to a couple of projects, I found that there are some additional features that I would like. This changeset adds a `select` function that takes a fetch action as an argument, so that I do not have to compute request params once in an action creator and again when selecting.

For example, if I have some action creators:

``` js
function fetchWidgets(page) {
  return fetchCollection(modelName, '/widgets', { page })
}

function fetchWidget(id) {
  return fetchRecord(modelName, id, `/widgets/${id}`)
}
```

I can reuse the action creators when selecting collections and records in components:

``` js
function mapStateToProps(state) {
  return {
    widgets: select(fetchWidgets(state.page), state.models),
    selectedWidget: select(fetchWidget(state.selectedId), state.models),
  }
}
```

I implemented `select` so that it includes the given action as a property of the selection object. This simplifies component initialization:

``` js
componentWillMount() {
  const { dispatch, widgets, selectedWidget } = this.props
  if (widgets.needsFetch) {
    dispatch(widgets.fetch)
  }
  if (selectedWidget.needsFetch) {
    dispatch(selectedWidget.fetch)
  }
}
```

I also included a `selectStatus` function that works like `selectNiceActionStatus`, but that takes a create/update/delete action as an argument instead of a model name. I am thinking of extending this to put transaction IDs in those actions (generally hidden from the user) that can be read by `selectStatus` to track the status of a specific request, even if it is not the latest request of a given type.

`select` and `selectStatus` by themselves were pretty small changes. I also set up Flow typechecking and added type definitions to `actionTypes`, `actionCreators`, and `selectors`. The advantage of that is that Flow can infer the type of data returned from a `select` call by referencing a type parameter on action creators. If I define my action creators with type annotations:

``` js
import type { CrudAction } from 'redux-crud-store/src'

function fetchWidgets(page: number): CrudAction<Widget[]> {
  return fetchCollection(modelName, '/widgets', { page })
}

function fetchWidget(id: string): CrudAction<Widget> {
  return fetchRecord(modelName, id, `/widgets/${id}`)
}
```

Flow will be able to determine via type inference that `widgets.data` is of type `Widget[]`, and that `selectedWidget.data` is of type `Widget`.

But if you do not want that Flow stuff in your library I can take it out, and send type definitions to the [flow-typed](https://github.com/flowtype/flow-typed) repository instead.

There are some other changes that I am thinking of submitting pull requests for:
- Add transaction IDs to create/update/delete actions to match status selections to requests more accurately
- Provide off-the-shelf API client implementation that uses `fetch` API
- Extend Flow integration for static checking of library internals (if that is appealing to you)
- Modify build system so that consumers who want typechecking do not have to import from 'redux-crud-store/src'
- Remove explicit dependency on 'babel-polyfill' (library consumers may not want this)
- Change `call` to `apply` when invoking API client

### CODE REVIEW

**Comment by devvmh** [comment]
This looks amazing. I love the type-flow stuff. I've never seen it before but it definitely makes the code more self-documenting, once you parse through it. I think that's a big strength for this library, because the API can be a bit arbitrary

I'm also digging the select function - looks great and it looks like it simplifies the component code even more, which is a huge plus.

I have a couple of  questions (but I would be fine to merge this anytime):
1. Can you write some documentation for the new functions, and maybe update the old documentation to describe the new ideal workflow? One implicit goal I've had is to make it possible to use this without diving too deep into the ES6 code, because of how hard it is to get into react/redux for the first time.
2. Would you like to be a collaborator on this repo + the npm project? (our company has paused our React development, so I'm guessing I won't be as active as you or I would like on this project).
3. Does any of this break the existing API? It doesn't look like it does to me, so I could just increment the minor version once I publish, but let me know if any changes aren't backwards compatible

Thanks a lot! I'm interested in all of the other changes you proposed. I think I'll open issues on them to discuss further.

File: src/sagas.js [hunk 1]
```diff
@@ -19,14 +19,16 @@ function* garbageCollector() {
...
-  action.meta.fetchTime = Date.now()
+  const meta = {
+    ...action.meta,
+    fetchTime: Date.now()
+  }
 
   try {
     const response = yield call(apiClient[method], path, { params, data })
-    yield put({ ...action, type: success, payload: response })
+    yield put({ ...meta, type: success, payload: response })
```
Comment (hallettj) [inline src/sagas.js]: Oh, this is wrong. I will make a note to fix it.

**Comment by hallettj** [comment]
Great, I'm glad you like these ideas.

In response to your questions:
1. Sure thing; documentation is no problem.
2. I would be happy to be a collaborator! I am working on a couple of React projects now, and I am angling for more React & React Native projects in the future. I can see using redux-crud-store in every case.
3. I took some care not to change the API (though I missed the accidental change pointed to by the line note above). But I do have code in a feature/transactionIds branch that changes the behavior of `selectStatus`, and that adds a map to the store. I would not minding publishing `select` quickly; but I would prefer not to introduce `selectStatus` right before potentially changing it.

I look forward to working on redux-crud-store-store with you!

**Comment by devvmh** [comment]
OK, it would be great if you could

1) update the redux-saga code so it doesn't break any existing functionality (we can make that change later)
2) update docs/API.md with the new functions

I could look into updating README.md once this is merged

Ping me once it's ready and I'll review again and merge

**Comment by hallettj** [comment]
Ok, I fixed the bug in the sagas, updated the documentation, and fixed an inconsistency in the return values for `select` (that was 65c1420).

I'm sorry, I read your post to quickly and missed the part about you wanting to update README.md yourself. Please treat my changes as suggestions.

I removed `selectStatus` for the moment. I would prefer to put in the version of `selectStatus` in my `feature/transactionIds` branch. That branch is code-complete, but needs to be merged with changes in this one, and needs updated documentation.

**Comment by devvmh** [comment]
No, this looks great, thanks for updating README.md. 

Flow annotations may take me a little while to get used to. It will be nice to use a real syntax in API.md, instead of one I made up.

I'll try to merge this alongside the other pull request tomorrow or next week, and publish a 4.0.0 after that. I'll deprecate selectActionStatus' old implementation at the same time, per #8.

**Comment by devvmh** [comment]
This is merged into a branch: https://github.com/uniqueway/redux-crud-store/tree/release/4.0.0. As I get time I will test this and then release + close this PR.

**Pull request closed without merging**